### PR TITLE
CI build for AMD Tensorflow & Pytorch

### PR DESCRIPTION
**Pull request by dibryant**
Fixes for https://issues.redhat.com/browse/RHOAIENG-6377
Contributing to PR: https://github.com/opendatahub-io/notebooks/pull/557
Adding CI builds for AMD Pytorch and Tensorflow.

### CODE REVIEW

**Review by atheo89 — CHANGES_REQUESTED** [review]
Added some review commends. Once you pass this PR from a second iteration of fixes you will need to run on your PC `make job` this will auto generate some files that will fix the complaining jobs

File: ci-operator/config/opendatahub-io/notebooks/opendatahub-io-notebooks-main.yaml [hunk 1]
```diff
@@ -160,7 +160,19 @@ images:
 - context_dir: jupyter/minimal/ubi9-python-3.9
   dockerfile_path: Dockerfile
   from: amd-c9s-python-3.9
-  to: amd-jupyter-minimal-c9s-python-3.9
+  to: amd-jupyter-minimal-ubi9-python-3.9
+- context_dir: jupyter/datascience/ubi9-python-3.9
+  dockerfile_path: Dockerfile
+  from: amd-jupyter-minimal-ubi9-python-3.9
+  to: amd-jupyter-datascience-ubi9-python-3.9
+- context_dir: jupyter/tensorflow/ubi9-python-3.9
```
Comment (harshad16) [inline ci-operator/config/opendatahub-io/notebooks/opendatahub-io-notebooks-main.yaml:168]:
```suggestion
- context_dir: jupyter/amd/tensorflow/ubi9-python-3.9
```

File: ci-operator/config/opendatahub-io/notebooks/opendatahub-io-notebooks-main.yaml [hunk 2]
```diff
@@ -160,7 +160,19 @@ images:
...
+  dockerfile_path: Dockerfile
+  from: amd-jupyter-datascience-ubi9-python-3.9
+  to: amd-jupyter-tensorflow-ubi9-python-3.9
+- context_dir: jupyter/pytorch/ubi9-python-3.9
```
Comment (harshad16) [inline ci-operator/config/opendatahub-io/notebooks/opendatahub-io-notebooks-main.yaml:172]:
```suggestion
- context_dir: jupyter/amd/pytorch/ubi9-python-3.9
```

File: ci-operator/config/opendatahub-io/notebooks/opendatahub-io-notebooks-main.yaml [hunk 3]
```diff
@@ -940,6 +952,65 @@ tests:
...
+- as: notebook-amd-jupyter-ds-ubi9-python-3-9-image-mirror
+  postsubmit: true
+  steps:
+    dependencies:
+      SOURCE_IMAGE_REF: amd-jupyter-datascience-ubi9-python-3.9
+    env:
+      IMAGE_REPO: workbench-images
+      IMAGE_TAG: YearIndex
+      RELEASE_VERSION: amd-jupyter-datascience-ubi9-python-3.9
+    workflow: opendatahub-io-ci-image-mirror
```
Comment (harshad16) [inline ci-operator/config/opendatahub-io/notebooks/opendatahub-io-notebooks-main.yaml:964]: For now, we can drop the AMD data-science and minimal, as we are not gonna provide them in.
leaving this for future.

Remove these lines, and the minimal as well, as we don't really provide them to the user.

File: ci-operator/config/opendatahub-io/notebooks/opendatahub-io-notebooks-main.yaml [hunk 4]
```diff
@@ -160,7 +160,19 @@ images:
 - context_dir: jupyter/minimal/ubi9-python-3.9
   dockerfile_path: Dockerfile
   from: amd-c9s-python-3.9
-  to: amd-jupyter-minimal-c9s-python-3.9
+  to: amd-jupyter-minimal-ubi9-python-3.9
```
Comment (harshad16) [inline ci-operator/config/opendatahub-io/notebooks/opendatahub-io-notebooks-main.yaml:163]:
```suggestion
  to: amd-jupyter-minimal-c9s-python-3.9
```


Even though we use the directory, ubi9-python-3.9
we are still basing these on centos , so we should use the centos in naming them
can you do this for the below ones as well.

**Review by harshad16 — CHANGES_REQUESTED** [review]
Thank you for the awesome work
some comments.

File: ci-operator/config/opendatahub-io/notebooks/opendatahub-io-notebooks-main.yaml [hunk 5]
```diff
@@ -157,10 +157,14 @@ images:
...
-- context_dir: jupyter/minimal/ubi9-python-3.9
+- context_dir: jupyter/amd/tensorflow/ubi9-python-3.9
+  dockerfile_path: Dockerfile
+  from: amd-jupyter-datascience-ubi9-python-3.9
+  to: amd-jupyter-tensorflow-c9s-python-3.9
+- context_dir: jupyter/amd/pytorch/ubi9-python-3.9
   dockerfile_path: Dockerfile
-  from: amd-c9s-python-3.9
-  to: amd-jupyter-minimal-c9s-python-3.9
+  from: amd-jupyter-datascience-ubi9-python-3.9
```
Comment (atheo89) [inline ci-operator/config/opendatahub-io/notebooks/opendatahub-io-notebooks-main.yaml:166]: This `amd-jupyter-datascience-ubi9-python-3.9` is not build (amd specific), I think you need here `jupyter-datascience-ubi9-python-3.9`

And again, once you apply the changes please run from your pc `make jobs` this command will fix some of the tests and also will autogenerate the post-submit jobs

**Comment by atheo89** [comment]
https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_release/53309/pull-ci-openshift-release-master-generated-config/1806328621720145920

```
ERROR: Run the following command to re-generate the Prow jobs:
ERROR: $ make jobs
```

File: ci-operator/config/opendatahub-io/notebooks/opendatahub-io-notebooks-main.yaml [hunk 6]
```diff
@@ -921,24 +925,45 @@ tests:
...
     dependencies:
-      SOURCE_IMAGE_REF: amd-jupyter-minimal-c9s-python-3.9
+      SOURCE_IMAGE_REF: amd-jupyter-tensorflow-c9s-python-3.9
+    env:
+      IMAGE_REPO: workbench-images
+      IMAGE_TAG: YearIndex
+      RELEASE_VERSION: amd-jupyter-tensorflow-c9s-python-3.9
+    workflow: opendatahub-io-ci-image-mirror
+- as: notebook-amd-jupyter-tf-c9s-python-3-9-pr-image-mirror
+  run_if_changed: (amd\/ubi9-python-3.9\/*)|(base\/ubi9-python-3.9\/*)|(jupyter\/minimal\/ubi9-python-3.9\/*)|(jupyter\/datascience\/ubi9-python-3.9\/*)|(jupyter\/amd\/tensorflow\/ubi9-python-3.9\/*)
```
Comment (atheo89) [inline ci-operator/config/opendatahub-io/notebooks/opendatahub-io-notebooks-main.yaml:939]:
```suggestion
  run_if_changed: (amd\/c9s-python-3.9\/*)|(jupyter\/datascience\/ubi9-python-3.9\/*)|(jupyter\/amd\/tensorflow\/ubi9-python-3.9\/*)
```

File: ci-operator/config/opendatahub-io/notebooks/opendatahub-io-notebooks-main.yaml [hunk 7]
```diff
@@ -921,24 +925,45 @@ tests:
...
-      SOURCE_IMAGE_REF: amd-jupyter-minimal-c9s-python-3.9
+      SOURCE_IMAGE_REF: amd-jupyter-pytorch-c9s-python-3.9
     env:
       IMAGE_REPO: workbench-images
-      RELEASE_VERSION: amd-jupyter-minimal-c9s-python-3.9
+      IMAGE_TAG: YearIndex
+      RELEASE_VERSION: amd-jupyter-pytorch-ubi9-python-3.9
+    workflow: opendatahub-io-ci-image-mirror
+- as: notebook-amd-jupyter-pyt-c9s-python-3-9-pr-image-mirror
+  run_if_changed: (amd\/ubi9-python-3.9\/*)|(base\/ubi9-python-3.9\/*)|(jupyter\/minimal\/ubi9-python-3.9\/*)|(jupyter\/datascience\/ubi9-python-3.9\/*)|(jupyter\/amd\/pytorch\/ubi9-python-3.9\/*)
```
Comment (atheo89) [inline ci-operator/config/opendatahub-io/notebooks/opendatahub-io-notebooks-main.yaml:959]:
```suggestion
  run_if_changed: (amd\/c9s-python-3.9\/*)|(jupyter\/datascience\/ubi9-python-3.9\/*)|(jupyter\/amd\/pytorch\/ubi9-python-3.9\/*)
```

File: ci-operator/config/opendatahub-io/notebooks/opendatahub-io-notebooks-main.yaml [hunk 8]
```diff
@@ -1497,6 +1522,29 @@ tests:
           cpu: 100m
           memory: 200Mi
     workflow: ipi-gcp
+- as: amd-notebooks-e2e-tests
+  run_if_changed: (base\/ubi9-python-3.9\/*)|(jupyter\/minimal\/ubi9-python-3.9\/*)|(jupyter\/datascience\/ubi9-python-3.9\/*)|(amd\/*)
```
Comment (atheo89) [inline ci-operator/config/opendatahub-io/notebooks/opendatahub-io-notebooks-main.yaml:1526]:
```suggestion
  run_if_changed: (amd\/c9s-python-3.9\/*)||(jupyter\/datascience\/ubi9-python-3.9\/*)|(jupyter\/amd\/pytorch\/ubi9-python-3.9\/*)||(jupyter\/amd\/tensorflow\/ubi9-python-3.9\/*)
```

File: ci-operator/config/opendatahub-io/notebooks/opendatahub-io-notebooks-main.yaml [hunk 9]
```diff
@@ -1497,6 +1522,29 @@ tests:
...
+  steps:
+    cluster_profile: gcp-opendatahub
+    test:
+    - as: amd-jupyter-ubi9-python-3.9-test-e2e
```
Comment (atheo89) [inline ci-operator/config/opendatahub-io/notebooks/opendatahub-io-notebooks-main.yaml:1530]:
```suggestion
    - as: amd-jupyter-tensorflow-python-3.9-test-e2e
```

File: ci-operator/config/opendatahub-io/notebooks/opendatahub-io-notebooks-main.yaml [hunk 10]
```diff
@@ -1497,6 +1522,29 @@ tests:
...
+      cli: latest
+      commands: |
+        oc new-project amd-jupyter-ubi9-python-3-9
```
Comment (atheo89) [inline ci-operator/config/opendatahub-io/notebooks/opendatahub-io-notebooks-main.yaml:1533]:
```suggestion
        oc new-project amd-jupyter-tensorflow-python-3-9
```

File: ci-operator/config/opendatahub-io/notebooks/opendatahub-io-notebooks-main.yaml [hunk 11]
```diff
@@ -1497,6 +1522,29 @@ tests:
...
+        IFS=':' read -r -a NOTEBOOK_IMAGE <<< "${JUPYTER_AMD_IMAGE}"
```
Comment (atheo89) [inline ci-operator/config/opendatahub-io/notebooks/opendatahub-io-notebooks-main.yaml:1534]:
```suggestion
        IFS=':' read -r -a NOTEBOOK_IMAGE <<< "${JUPYTER_AMD__TENSORFLOW_IMAGE}"
```

File: ci-operator/config/opendatahub-io/notebooks/opendatahub-io-notebooks-main.yaml [hunk 12]
```diff
@@ -1497,6 +1522,29 @@ tests:
...
+        make deploy8-amd-ubi9-python-3.9 \
```
Comment (atheo89) [inline ci-operator/config/opendatahub-io/notebooks/opendatahub-io-notebooks-main.yaml:1535]:
```suggestion
        make deploy9-amd-ubi9-python-3.9 \
```
Not tested, but to be sure please check it localy with the makefile

File: ci-operator/config/opendatahub-io/notebooks/opendatahub-io-notebooks-main.yaml [hunk 13]
```diff
@@ -1497,6 +1522,29 @@ tests:
...
+            -e IMAGE_REGISTRY="${NOTEBOOK_IMAGE[0]}" -e NOTEBOOK_TAG="${NOTEBOOK_IMAGE[1]}"
+        make test-amd-ubi9-python-3.9
+        make undeploy8-amd-ubi9-python-3.9
```
Comment (atheo89) [inline ci-operator/config/opendatahub-io/notebooks/opendatahub-io-notebooks-main.yaml:1538]:
```suggestion
        make undeploy9-amd-ubi9-python-3.9
```

File: ci-operator/config/opendatahub-io/notebooks/opendatahub-io-notebooks-main.yaml [hunk 14]
```diff
@@ -1497,6 +1522,29 @@ tests:
...
+      dependencies:
+      - env: JUPYTER_AMD_IMAGE
```
Comment (atheo89) [inline ci-operator/config/opendatahub-io/notebooks/opendatahub-io-notebooks-main.yaml:1540]:
```suggestion
      - env: JUPYTER_AMD__TENSORFLOW_IMAGE
```

File: ci-operator/config/opendatahub-io/notebooks/opendatahub-io-notebooks-main.yaml [hunk 15]
```diff
@@ -1497,6 +1522,29 @@ tests:
...
+        name: amd-jupyter-ubi9-python-3.9
```
Comment (atheo89) [inline ci-operator/config/opendatahub-io/notebooks/opendatahub-io-notebooks-main.yaml:1541]:
```suggestion
        name: amd-jupyter-tensorflow-python-3.9
```

File: ci-operator/config/opendatahub-io/notebooks/opendatahub-io-notebooks-main.yaml [hunk 16]
```diff
@@ -1497,6 +1522,29 @@ tests:
           cpu: 100m
           memory: 200Mi
     workflow: ipi-gcp
+- as: amd-notebooks-e2e-tests
```
Comment (atheo89) [inline ci-operator/config/opendatahub-io/notebooks/opendatahub-io-notebooks-main.yaml:1525]: Right now, we have test only for tensorflow, you will need to do the same for pytorch .
Hint: copy the lines from 1530-1546 and replace it accordingly

**Review by atheo89 — CHANGES_REQUESTED** [review]
New iteration of comments

**Comment by harshad16** [comment]
/hold
reviewing

File: ci-operator/config/opendatahub-io/notebooks/opendatahub-io-notebooks-main.yaml [hunk 17]
```diff
@@ -157,10 +157,14 @@ images:
   dockerfile_path: Dockerfile
   from: base-c9s-python-3.9
   to: amd-c9s-python-3.9
-- context_dir: jupyter/minimal/ubi9-python-3.9
+- context_dir: jupyter/amd/tensorflow/ubi9-python-3.9
+  dockerfile_path: Dockerfile
+  from: jupyter-datascience-ubi9-python-3.9
+  to: amd-jupyter-tensorflow-c9s-python-3.9
+- context_dir: jupyter/amd/pytorch/ubi9-python-3.9
```
Comment (harshad16) [inline ci-operator/config/opendatahub-io/notebooks/opendatahub-io-notebooks-main.yaml:164]: We need to build minimal,datascience with AMD
however we don't push them to quay.io, they would only be used for building 
pytorch and tensorflow.

```suggestion
- context_dir: jupyter/minimal/ubi9-python-3.9
  dockerfile_path: Dockerfile
  from: amd-c9s-python-3.9
  to: amd-jupyter-minimal-c9s-python-3.9
- context_dir: jupyter/data-science/ubi9-python-3.9
  dockerfile_path: Dockerfile
  from: amd-jupyter-minimal-c9s-python-3.9
  to: amd-jupyter-datascience-c9s-python-3.9
- context_dir: jupyter/amd/tensorflow/ubi9-python-3.9
  dockerfile_path: Dockerfile
  from: amd-jupyter-datascience-c9s-python-3.9
  to: amd-jupyter-tensorflow-c9s-python-3.9
- context_dir: jupyter/amd/pytorch/ubi9-python-3.9
```



Even though we are using c9s base image, we use the structure of ubi9-python3.9 in others
as there is no OS based information.

**Review by harshad16 — CHANGES_REQUESTED** [review]
Require following changes for the build of AMD based TF and PyTorch

**Comment by atheo89** [comment]
/pj-rehearse pull-ci-opendatahub-io-notebooks-main-notebook-rocm-c9s-python-3-9-pr-image-mirror pull-ci-opendatahub-io-notebooks-main-notebook-rocm-jupyter-pyt-c9s-python-3-9-pr-image-mirror pull-ci-opendatahub-io-notebooks-main-notebook-rocm-jupyter-tf-c9s-python-3-9-pr-image-mirror pull-ci-opendatahub-io-notebooks-main-rocm-notebooks-e2e-tests

**Comment by atheo89** [comment]
@dibryant PTAL, Once this pull request https://github.com/opendatahub-io/notebooks/pull/609 is merged, it will fix the failed test for the TensorFlow image build.

**Comment by jiridanek** [comment]
The right command should actually be
/pj-rehearse pull-ci-opendatahub-io-notebooks-main-notebook-rocm-jupyter-tf-c9s-python-3-9-pr-image-mirror

**Comment by jiridanek** [comment]
ah, ok, so rebase needs to be done first, then we can rehearse, and then hopefully this will work

**Review by harshad16 — APPROVED** [review]
/lgtm
/approve
/unhold
/pj-rehearse ack
thanks :+1:

**Comment by harshad16** [comment]
/lgtm
/approve
/pj-rehearse ack
thanks 👍

**Pull request merged**